app-marketplace-v2-spec/architecture.md

```diff
@@ -0,0 +1,146 @@
+respond as either:
+  - 200 OK + (optional) simple markdown + (optional) JSON.
+  - Error code + error message.
+  - A (302?) "redirect" to navigate to a channel, external browser, an iframe
+    popup, or run another function.
+  - #TODO @larkox to reconcile with the "side effects" of his Integrations 2.0
+    spec.
+  - #TODO @levb, @larkox to propose redirect and execute response format.
+  - #TODO @crspeller, @lieut-data anything to borrow/learn from the workflow
+    projects.
+
+### Mattermost to Cloud App.
+Mattermost interacts with Cloud Apps in 3 ways: 
+  - **Functions** that are invoked via Interactive Inputs, and have
+    request/response semantics. They may be synchronous, or execute in the
+    background once invoked.
+  - 1-way **Change notifications** that are sent after changes occur (from the
+    server).
+  - **Autocomplete** queries.
+
+### Functions 
+1. Functions are fundamental units of executing specific user instructions
+by Cloud Apps. 
+2. They are called by Slash Commands, Widgets, Modal and Bot Dialogs. Commands, Modal and Bot dialogs are also callable.
+3. Functions may accept both structured JSON inputs, as well as "raw" command strings with options.
+4. Dialogs and commands that are fully valid against SlashCommands definition are expected to submit structured data.
+5. Widgets "hardcode" their invocation of functions, so they can do as they please.
+6. Slash Commands invoke the closest-matched function. If the command string passes validation, it is submitted as JSON, otherwise as raw text. Raw text is always submitted.
+7. Expandable context provides functions with most data dependencies pre-fetched, as instructed by the App Manifest, or the function binding.
+
+### Interactive Inputs.
+See [Interactive Inputs](commands.md) for more.
+  - **Slash Commands** with autocomplete, execute functions.
+  - **Widgets** that can be bound to directly execute functions (Post menu items, channel menu, interactive post buttons, etc). In the future they may be extended to become in-place forms or autocomplete boxes.
+  - **Modal Dialogs** and **Bot Dialogs**, that gather/edit parameter values, then execute a function.
+
+#### Slash Commands
+1. Slash commands will have interactive auto-complete, available in the usual message box in Mattermost. All declared commands that are not "hidden" will be available for autocomplete.
+2. #TODO we need to be able to re-define what is visible (valid?) based on the state already entered, and hide the irrelevant options. @iomodo can you take a look at this?
+3. #STRETCH a "slash command" widget that would execute a pre-configured autocomplete command, in the given context (team/channel or Bot DM), and display the result.
+4. Slash commands may have a --modal option that would launch a modal to gather the inputs, rather than autocomplete. The parameters in the slash command are used as the default values in the modal.
+5. All valid slash commands must be declared in the App manifest
```

Comment:
Can slash commands be dynamically added/modified by the App?

Reply:
No, but the "hidden" and "required" can be updated mid-entry. Just like the Modal Dialogs. Eventually we can add REST API calls to modify the definitions.